### PR TITLE
Rhi implementation

### PR DESCRIPTION
**Pull request by blakdragan7**
**This is an example for Chris to see the difference in RHI. Please do not merge.**

Chris the files you should look at
*  `Source\RiveRenderer\Private\Platform\pls_render_context_rhi_impl.hpp|cpp`
*  `Source\RiveRenderer\Private\Platform\RiveRenderTargetRHI.h|cpp`
*  `Source\RiveRenderer\Private\Platform\RiveRendererRHI.h|cpp`
* `Source\RiveRenderer\Private\Shaders\ShaderPipelineManager.h|cpp`

If you want to look at the shader code here you can but that will also be in mono. The shader code path is

`Shaders\Private\Rive`

### CODE REVIEW

**Review by csmartdalton — COMMENTED** [review]
Wow, what a huge drop! 😁 This is really, really high quality code. Awesome work hooking this all up Jonathon!!

I've scattered around a bunch of comments that would be nice to think through. Let me know if you want to discuss any of them and/or have other thoughts.

File: Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+template<typename UniformBufferType>
+class UniformBufferRHIImpl : public rive::gpu::BufferRing
+{
+public:
+    UniformBufferRHIImpl(size_t in_sizeInBytes): BufferRing(in_sizeInBytes)
+    {
+        //m_buffer =  TUniformBufferRef<UniformBufferType>::CreateEmptyUniformBufferImmediate( UniformBuffer_MultiFrame);
+    }
+
+    void Sync(FRHICommandList& commandList, int offset)
```
Comment (csmartdalton) [inline Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.hpp:73]: `offset` should probably be a `size_t` if dealing when memory sizes.

And just for my own understanding, Is the size built into `UniformBuffer_SingleFrame`? How does this know how much data to sync?

File: Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,1207 @@
...
+        const size_t size = ResourceArray.GetResourceDataSize();
+        FRHIResourceCreateInfo Info(DebugName, &ResourceArray);
+        auto buffer = RHICmdList.CreateBuffer(size,
+            EBufferUsageFlags::Static | bindFlags,sizeof(DataType),
+            ERHIAccess::VertexOrIndexBuffer, Info);
+        return buffer;
+    }
+    
+#define SYNC_BUFFER(buffer, command_list) if(buffer)buffer->Sync(command_list);
+#define SYNC_BUFFER_WITH_OFFSET(buffer, command_list, offset)if(buffer)buffer->Sync(command_list, offset);
```
[the remark this answers is not in the document]
Comment (csmartdalton) [reply Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.cpp:170]: Actually this looks like it's only used for uniform buffers, right? Could it just be called `SYNC_UNIFORM_BUFFER` or something?

File: Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,1207 @@
...
+    GraphicsPSOInit.BoundShaderState.VertexShaderRHI = VSShader.GetVertexShader();
+    GraphicsPSOInit.BoundShaderState.PixelShaderRHI = PSShader.GetPixelShader();
+    SetGraphicsPipelineState(CommandList, GraphicsPSOInit, 0, EApplyRendertargetOption::CheckApply, true, EPSOPrecacheResult::NotSupported);
+}
+
+template<typename ShaderType>
+void SetParameters(FRHICommandList& CommandList, FRHIBatchedShaderParameters& BatchedParameters,
+    TShaderMapRef<ShaderType> Shader, typename ShaderType::FParameters& VParameters)
+{
+    ClearUnusedGraphResources(Shader, &VParameters);
```
[the remark this answers is not in the document]
Comment (blakdragan7) [reply Source/RiveRenderer/Private/Platform/pls_render_context_rhi_impl.cpp:54]: Its convention for unreal, but it allows permutations to not attempt to bind to something that doesnt exists.

**Pull request closed without merging**